### ISSUE TITLE
Tracking: Unstable Rust feature gates used by Servo

### ISSUE THREAD

**Issue by Ms2ger** [servo/servo#5286]
Servo uses some unstable Rust features.

## Required

* Some have a clear path to stabilization and are being worked on:
  - [ ] const_fn (rust-lang/rust#24111)
    Note that `min_const_fn` is not sufficient for us: https://github.com/rust-lang/rust/issues/53555#issuecomment-418755997
  - [ ] register_tool (https://github.com/rust-lang/rust/issues/66079)
* Others are more problematic. They’re not going away soon, but on the long term we may want to find alternatives:
  - Compiler plugin for a custom lint (with whitelisted attributes). Could this be built on top of RLS?
    - [ ] plugin, plugin_registrar (rust-lang/rust#29597)
    - [ ] rustc_private (rust-lang/rust#27812)

## Optional

* Optimizations in crates in other repositories
  - `parking_lot`
    - [ ] integer_atomics (https://github.com/rust-lang/rust/issues/32976)
    - [ ] asm (https://github.com/rust-lang/rust/issues/29722)
* Micro-benchmarks
  - [ ] test (rust-lang/rust#27812)
* Diagnostics / debugging
  - [ ] on_unimplemented (rust-lang/rust#29628)
  - [ ] core_intrinsics: , `breakpoint`, `abort`

## No longer used, or have been stabilized

- [x] alloc (rust-lang/rust#27783)
- [x] allocator_api (rust-lang/rust#32838)
- [x] append (rust-lang/rust#27765)
- [x] arc_unique (rust-lang/rust#27718)
- [x] arc_weak (rust-lang/rust#27718)
- [x] as_unsafe_cell (rust-lang/rust#27708)
- [x] ascii_ctype (rust-lang/rust#39658)
- [x] borrow_state (rust-lang/rust#27733)
- [x] box_patterns (rust-lang/rust#29641)
- [x] box_raw (rust-lang/rust#27768)
- [x] box_str (rust-lang/rust#27785)
- [x] box_syntax (rust-lang/rust#27779)
- [x] cfg_target_feature (rust-lang/rust#29717)
- [x] collections
- [x] conservative_impl_trait (rust-lang/rust#34511)
- [x] convert (rust-lang/rust#27729)
- [x] core (rust-lang/rust#27701)
- [x] custom_attribute
- [x] custom_derive
- [x] drain (rust-lang/rust#27711)
- [x] env
- [x] exit_status
- [x] filling_drop (rust-lang/rust#5016)
- [x] float_consts
- [x] fn_box (rust-lang/rust#28796)
- [x] fn_must_use (https://github.com/rust-lang/rust/issues/43302)
- [x] fs_walk (rust-lang/rust#27707)
- [x] get_type_id (rust-lang/rust#27745)
- [x] global_allocator (rust-lang/rust#27389)
- [x] hash
- [x] hasher_write
- [x] hashmap_hasher (rust-lang/rust#27713)
- [x] heap_api (rust-lang/rust#27700)
- [x] int_uint
- [x] io
- [x] ip_addr (rust-lang/rust#27801)
- [x] iter_arith (rust-lang/rust#27739)
- [x] iter_cmp (rust-lang/rust#27724)
- [x] iterator_step_by (rust-lang/rust#27741)
- [x] libc
- [x] link_args (rust-lang/#29596)
- [x] mpsc_select (rust-lang/rust#27800)
- [x] negate_unsigned
- [x] net
- [x] nonzero (rust-lang/rust#27730)
- [x] old_io
- [x] old_path
- [x] oom (rust-lang/rust#27700)
- [x] optin_builtin_traits (rust-lang/rust#13231)
- [x] os
- [x] owned_ascii_ext
- [x] page_size
- [x] path
- [x] path_ext (rust-lang/rust#27725)
- [x] platform_intrinsics (rust-lang/rust#27731)
- [x] quote
- [x] range_contains (rust-lang/rust#32311)
- [x] rc_unique (rust-lang/rust#27718)
- [x] reflect_marker (rust-lang/rust#27749)
- [x] repr_simd (rust-lang/rust#27731)
- [x] result_expect (rust-lang/rust#27277)
- [x] slice_bytes (rust-lang/rust#27740)
- [x] slice_chars
- [x] slice_extras
- [x] slice_patterns (rust-lang/rust#23121)
- [x] slice_position_elem
- [x] start
- [x] std_misc
- [x] stdsimd (rust-lang/rust#27731)
- [x] step_trait (rust-lang/rust#27741)
- [x] stmt_expr_attributes (rust-lang/rust#15701)
- [x] str_char (rust-lang/rust#27754)
- [x] str_split_at (rust-lang/rust#27792)
- [x] str_utf16 (rust-lang/rust#27714)
- [x] str_words
- [x] struct_field_attributes (rust-lang/rust#41681)
- [x] thread_local
- [x] thunk
- [x] try_from (rust-lang/rust#33417)
- [x] type_name (https://github.com/rust-lang/rfcs/issues/1428)
- [x] unboxed_closures
- [x] unicode (rust-lang/rust#27784)
- [x] unsafe_destructor
- [x] unsafe_no_drop_flag
- [x] unique (rust-lang/rust#27730)
- [x] untagged_unions (rust-lang/rust#32836)
- [x] use_extern_macros (rust-lang/rust#35896)
- [x] vec_from_raw_buf
- [x] vec_push_all (rust-lang/rust#27744)
- [x] zero_one (rust-lang/rust#27739)

**Comment by SimonSapin** [comment]
I believe `env` is stable with the latest Rust upgrade.

**Comment by SimonSapin** [comment]
So is `hash`.

**Comment by SimonSapin** [comment]
The `int_uint` feature does not exist anymore in Nightly, using `int` or `uint` types, or `u` or `i` suffixes is now always an error. Let’s get rid of this one before the next rustup. (Simply use `usize` and `isize` types, and remove suffixes if possible or use `_usize` or `_isize`).

**Comment by Ms2ger** [comment]
https://gist.github.com/Manishearth/967bc78db5c751c38d03

**Comment by bluss** [comment]
I'd avoid slice_patterns since it has so many open memory safety bugs.

**Comment by mbrubeck** [comment]
I updated the list above, and submitted #16681 and #16680 to remove some unnecessary feature uses.

Now the only uses of the `core_intrinsics` feature are for `breakpoint` (used in `script::dom::window::WindowMethods::Trap`) and `type_name` (used in debug-only output in `JS::trace`. Should we propose stable wrappers for these?

**Comment by SimonSapin** [comment]
I think the list above was created / is maintained by grepping this repository, so I also had a look at dependencies. 

**TL;DR**: there’s stuff that’s being removed, stuff we can disable, and a couple things in `rust-mozjs`.

----

To do so, I ran `cargo +stable build -j200` repeatedly in the `ports/servo` directory. The tries to build servo with a stable compile which quickly fails, but makes Cargo schedule as many jobs as the dependency graph allows before that happens. With enough repetition, I *think* this compiles every crate that can be compiled, leaving only `Compiling foo…` lines only for those that cause build errors. I then looked at those that were crates.io or git dependencies (i.e. everything not in this repository).

* Some crates had a Cargo feature flag named `unstable` or `nightly` that enable optimizations but apparently not library features that we require:
  * In `euclid` it uses `asm` and `repr_simd` features to enable a `SideOffsets2DSimdI32` type… that we don’t use, as far as I can tell.
  * In `tendril` (through `html5ever` and `xml5ever`) it uses the `nonzero` feature to enables `Option<Tendril<…>>` memory layout optimization.
  * In `parking_lot` it uses `integer_atomics` and `asm` for [optimizations](https://github.com/Amanieu/parking_lot/#nightly-vs-stable) (and `const_fn` to make some `fn`s be `const`, but we don’t rely on that).
* `multistr` uses `never_type`. That can be replaced with the `void` crate: https://github.com/clarcharr/multistr/pull/2
* `parse-hosts` uses `type_ascription`. That can be replaced with specifying a type parameter in a method call. (I’ll send a PR once the `multistr` one lands).
* `js` a.k.a. `servo/rust-mozjs` uses `const_fn`, `nonzero`, and `link_args`.
  * It looks like we do rely on at least one constructor being `const`, so not much to do here besides nudging https://github.com/rust-lang/rust/issues/24111
  * I think `NonZero` can be replaced with `std::ptr::Shared`, which might be stabilized Soon<sup>®</sup>: https://github.com/rust-lang/rust/issues/27730#issuecomment-300530242
  * The crate still builds with `#![feature(link_args)]` commented out. Do we not use it anymore? Was it stabilized?

I *think* that’s everything. I’ve update the original comment for unsolved stuff.

**Comment by SimonSapin** [comment]
For `const_fn` in rust-mozjs we could in theory expose a macro instead of a `const fn`, but as far as I can tell these `new_bitfield_*` constructors are generated by rust-bindgen. For `nonzero` we could have a Cargo feature flag like Tendril to use a no-op wrapper struct instead of `NonZero` when disabled, but this may not be worth the bother.

**Comment by SimonSapin** [comment]
With https://github.com/servo/servo/pull/18854 it’s now possible to disable the use of *some* unstable features with `./mach build --no-default-features --features default-except-unstable`

**Comment by jadbox** [comment]
What is Servo doing with/out box_patterns for match statements?

**Comment by jdm** [comment]
@jadbox We removed our use of that feature.

**Comment by SimonSapin** [comment]
@LifeIsStrange Yes, but `std::arch` contains individual architecture-specific instructions whereas `components/gfx/text/glyph.rs` uses the portable `u32x4` type. Using the former is a lot more involved. The latter is under discussion to be stabilized in the standard library eventually https://github.com/rust-lang/rfcs/pull/2366.

**Comment by SimonSapin** [comment]
Closing since this is out of date and I’m not convinced taking the time to keep it up to date is all that valuable. We can revisit after the main blocker (compiler plugins) is solved.
